crates/algorithms/src/haversine/mod.rs

```diff
@@ -0,0 +1,131 @@
+use core::fmt::Debug;
+use nalgebra::Point2;
+use num_traits::Float;
+
+///calculate distance between point with use haversine formula
+/// return distance in km
+#[cfg_attr(
+    feature = "tracing",
+    tracing::instrument("Calculate Haversine Data", skip_all)
+)]
+fn calculate_haversine_distance<T>(p1: Point2<T>, p2: Point2<T>, sphere_radius: T) -> T
+where
+    T: Copy + Float + Debug + 'static,
+{
+    //define two = 2 to support generic
+    let two = T::from(2.0).unwrap();
+    //calc delta
+    let dlat = (p2.x - p1.x).to_radians();
+    let dlon = (p2.y - p1.y).to_radians();
+    //convert coordinate to radians
+    let lat1_radians = p1.x.to_radians();
+    let lat2_radians = p2.x.to_radians();
+    //apply haversine formula
+    let a = (dlat / two).sin().powi(2)
+        + (dlon / two).sin().powi(2) * lat1_radians.cos() * lat2_radians.cos();
+    //use atan2 instead of arcsin(\sqrt a) due to numerical stability and precision
+    let c = two * a.sqrt().atan2((T::one() - a).sqrt());
+
+    sphere_radius * c
+}
+/// Returns the bearing from the first point to the second point.
+#[cfg_attr(
+    feature = "tracing",
+    tracing::instrument("Calculate Bearing Between Points", skip_all)
+)]
+fn return_bearing_between_points<T>(p1: Point2<T>, p2: Point2<T>) -> T
+where
+    T: Copy + Float + Debug + 'static,
+{
+    let lat1_rad = p1.x.to_radians();
+    let lat2_rad = p2.x.to_radians();
+    let lon_diff_rad = (p2.y - p1.y).to_radians();
+    let y = lon_diff_rad.sin() * lat2_rad.cos();
+    let x = lat1_rad.cos() * lat2_rad.sin() - lat1_rad.sin() * lat2_rad.cos() * lon_diff_rad.cos();
+
+    let bearing_rad = y.atan2(x);
+    //convert to degree because navigation work with degree and return it
+    let circle_angle = T::from(360.0).unwrap();
+    (bearing_rad.to_degrees() + circle_angle) % circle_angle
+}
+#[cfg(feature = "pregenerated")]
+macro_rules! impl_p_i_p_algorithm {
```

Comment:
this is not a point_in_polygon_algorithm, so I think this can safely be renamed impl_harversine_formula^^